git-record/src/ui.rs

```diff
@@ -0,0 +1,1041 @@
+use cursive::theme::{BaseColor, Effect};
+use cursive::traits::{Nameable, Resizable};
+use cursive::utils::markup::StyledString;
+use cursive::views::{Checkbox, Dialog, HideableView, LinearLayout, ScrollView, TextView};
+use cursive::{CursiveRunnable, CursiveRunner, View};
+use tracing::error;
+
+use crate::cursive_utils::{EventDrivenCursiveApp, EventDrivenCursiveAppExt};
+use crate::tristate::{Tristate, TristateBox};
+use crate::{FileHunks, Hunk, HunkChangedLine, RecordError, RecordState};
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum HunkType {
+    Before,
+    After,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct FileKey {
+    file_num: usize,
+}
+
+impl Display for FileKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self { file_num } = self;
+        write!(f, "FileKey({})", file_num)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct HunkKey {
+    file_num: usize,
+    hunk_num: usize,
+}
+
+impl Display for HunkKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self { file_num, hunk_num } = self;
+        write!(f, "HunkKey({},{})", *file_num, *hunk_num)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct HunkLineKey {
+    file_num: usize,
+    hunk_num: usize,
+    hunk_type: HunkType,
+    hunk_line_num: usize,
+}
+
+impl Display for HunkLineKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let Self {
+            file_num,
+            hunk_num,
+            hunk_type,
+            hunk_line_num,
+        } = self;
+        write!(
+            f,
+            "HunkLineKey({},{},{},{})",
+            file_num,
+            hunk_num,
+            match hunk_type {
+                HunkType::Before => "B",
+                HunkType::After => "A",
+            },
+            hunk_line_num
+        )
+    }
+}
+
+/// UI component to record the user's changes.
+pub struct Recorder {
+    did_user_confirm_exit: bool,
+    state: RecordState,
+}
+
+impl Recorder {
+    /// Constructor.
+    pub fn new(state: RecordState) -> Self {
+        Self {
+            did_user_confirm_exit: false,
+            state,
+        }
+    }
+
+    /// Run the terminal user interface and have the user interactively select
+    /// changes.
+    pub fn run(self, siv: CursiveRunner<CursiveRunnable>) -> Result<RecordState, RecordError> {
+        EventDrivenCursiveAppExt::run(self, siv)
+    }
+
+    fn make_main_view(&self, main_tx: Sender<Message>) -> impl View {
+        let mut view = LinearLayout::vertical();
+
+        let RecordState { files } = &self.state;
+
+        let count_changed_hunks = |&FileHunks { ref hunks }| {
+            hunks
+                .iter()
+                .filter(|hunk| match hunk {
+                    Hunk::Unchanged { .. } => false,
+                    Hunk::Changed { .. } => true,
+                })
```

Comment:
I think you can use `matches!(hunk, Changed { .. })` instead of the `match hunk { ... }` if you like

Reply:
In this case, I want to preserve exhaustiveness-checking.